### PR TITLE
[C#] Implemented Concept exercise: equality

### PR DESCRIPTION
**Pull request by mikedamay**
Please review design.md, instructions.md Example.cs and Equality.cs.

The `Equals()` method may lead to many representations.

### CODE REVIEW

**Comment by mikedamay** [comment]
I will reissue after some changes.

File: languages/csharp/exercises/concept/equality/.docs/introduction.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+- An overridden `Equals()` will contain equality tests on members of simple types using `==` and reference types with recursive calls to `Equals()`.
+- The static method `object.ReferenceEquals()` is used to compare two objects to detect if they are one and the same instance.
+
+### `Object.GetHashCode()`
+
+- `object.GetHashCode()` returns a hash code in the form of a 32 bit integer. The hash code is used by dictionary and set classes such as `Dictionary<T>` and `HashSet<T>`to store and retrieve objects in a performant manner.
+- The relationship between hash code and equality is that if two objects are equal (`Equal()` returns true) then `GetHashCode()` for the two objects must return the same value. This does not apply in the reverse direction. It is not symmetrical. Picture a lookup function that first goes to a "bucket" based on the hash code and then picks out the particular item using the equality test.
+- The easiest way to create a hashcode is to call `HashCode.Combine()` passing in the values used in the equality test (or a subset). Bear in mind the more information you provide to `Combine()` the more performant the hash implementation is likely to be.
```
[the remark this answers is not in the document]
Comment (mikedamay) [reply languages/csharp/exercises/concept/equality/.docs/introduction.md:15]: What about in the intro "The `HashCode` library API discusses the best way to generate a hash code"?

**Comment by mikedamay** [comment]
I have not seriously addressed the review points except https://github.com/exercism/v3/pull/1691#discussion_r445481510.  I'm just making sure it's OK to reuse the story for `object-initialization`.  I think it is.

File: languages/csharp/exercises/concept/equality/.docs/instructions.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+Implement `Authenticator.IsRegistered()` and ensure it returns false when no identities have been registered.
+
+```csharp
+var authenticator = new Authenticator();
+authenticator.IsRegistered(new Identity("alice@thecompetition.com", new FacialFeatures("blue", 0.8m)));
+// => false
+```
+
+## 3. Register new identities
```
[the remark this answers is not in the document]
Comment (mikedamay) [reply languages/csharp/exercises/concept/equality/.docs/instructions.md:33]: It was a deliberate decision but the reasoning escapes me.  Now reversed.

File: languages/csharp/exercises/concept/equality/.docs/introduction.md [hunk 3]
```diff
@@ -0,0 +1,16 @@
+This exercise looks at how you test for equality and the importance of hash codes.
```
[the remark this answers is not in the document]
Comment (mikedamay) [reply languages/csharp/exercises/concept/equality/.docs/introduction.md:1]: It was intentional (my latest approach is to do _after.md_ first and then cherry-pick) but I always favor consistency, all else being equal so I will have amended it.

**Comment by mikedamay** [comment]
Closed for proofing - I will reopen shortly

**Comment by mikedamay** [comment]
I would have liked to started out with the identity and facial recognition classes being mutable and invited to the student to consider this what changes they would make when adding identity but I was concerned about being pulled up by the refactoring police.  Any thoughts?

File: languages/csharp/exercises/concept/equality/.docs/instructions.md [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+In all occurrences the eye color parameter is guaranteed to be non-null.
+
+## 1. Authenticate the system administrator
+
+Despite your protests the system administrator insists on having a built-in identity during acceptance testing so that they can always be authenticated.
+
+The admin's email is admin@exerc.ism. They have green eyes and a philtrum with a width of 0.9.
+
+Add equality routines for the `Identity` and `FacialFeatures` classes.
```
[the remark this answers is not in the document]
Comment (mikedamay) [reply languages/csharp/exercises/concept/equality/.docs/instructions.md]: I tried something

File: languages/csharp/exercises/concept/equality/.docs/introduction.md [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+- Simple types (strings and primitives) are typically tested for equality with the `==` and `!=`.
+- Reference types (Instances of classes) are compared using the `Equals()` method inherited from `object`. If your goal with the equality test is to ensure that two objects are the exact same instance then relying on `object`'s implementation will suffice. If not, you need to override `object.Equals()`.
+- If you know that all the instances of your class are created in one place, say characters in some game or simulation then reference equality is sufficient. However, it is likely that multiple instances of the same real-world entity will be created (from a database, by user input, via a web request). In this case values that uniquely identify the entity must be tested for equality. Therefore `Equals()` must be overridden and appropriate data members of your class are tested for equality.
+- An overridden `Equals()` will contain equality tests on members of simple types using `==` and reference types with recursive calls to `Equals()`.
+- The static method `object.ReferenceEquals()` is used to compare two objects to detect if they are one and the same instance.
```
[the remark this answers is not in the document]
Comment (mikedamay) [reply languages/csharp/exercises/concept/equality/.docs/introduction.md]: I have cunningly hidden the goodies with "The Object class provides appropriate methods to compare two objects to detect if they are one and the same instance."

File: languages/csharp/exercises/concept/equality/.docs/introduction.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (ErikSchierboom) [reply languages/csharp/exercises/concept/equality/.docs/introduction.md:1]: Great!

**Pull request merged**